File: Interview-Pratice-FANNG/2-D-Array/bfs.js
const data = require('./data')
const directions = require('./directions')

// function BFS(matrix, row, column, values, explored,queue) {
//      // Check if two max 
//     if (row < 0 || column < 0 || row >= matrix.length || column >= matrix[0].length || explored[row][column]) return;
//     explored[row][column] = true;
//     values.push(matrix[row][column])
//     queue.push({row,column})
//     while(queue.length) {
//         let val = queue.shift()
//         for(let i = 0 ;i < directions.length ; i ++) {
//             BFS(matrix,val.row + directions[i][0], val.column +directions[i][1],values,explored,queue)
//         }
//     }

// }

const BFS_2_D_Array = (matrix) => {
    const explored = new Array(matrix.length).fill(0).map(el => new Array(matrix[0].length).fill(false))
    const values = []
    const queue = []
    queue.push({row:2,column:2})
    while(queue.length) {
        let firstQueue = queue.shift()
        let {row,column} = firstQueue
        if (row < 0 || column < 0 || row >= matrix.length || column >= matrix[0].length || explored[row][column]) {
            continue;
        }

        explored[row][column] = true;
        values.push(matrix[row][column])
        for(let i = 0 ; i < directions.length ;  i ++) {
            queue.push({row: row+directions[i][0],
                        column: column + directions[i][1]})
        }
    }
    return values
    
}
console.log(BFS_2_D_Array(data))

File: Interview-Pratice-FANNG/2-D-Array/bfs.py
from data import data
from directions import directions

def bfs(data):
    seen = [[False for i in range(len(data[0]))] for j in range(len(data))]
    queue = [[0,0]]
    values = []
    while(len(queue)):
        currentVal = queue.pop(0)
        row = currentVal[0]
        column = currentVal[1]
        if row < 0 or column < 0 or row >= len(data) or column >= len(data[0]) or seen[row][column]:
            continue
        seen[row][column] = True
        values.append(data[row][column])
        for direction in directions :
            next_row = row+ direction[0]
            next_col = column + direction[1]
            queue.append([next_row,next_col])
    return values
print(bfs(data))
